lebedigital/calibration/utils.py

```diff
@@ -0,0 +1,195 @@
+    if KG is not True:
+        dia = 98.6
+        height = 300.2
+
+    df['displacement'] = (df['Transducer 1[mm]'] + df['Transducer 2[mm]'] + df['Transducer 3[mm]']) / 3
+    df['stress'] = df['Force [kN]'] / (np.pi * (dia / 2) ** 2)
+
+    output = {
+        'height': height,
+        'diameter': dia,
+        'displacement': np.array(df['displacement']),
+        'stress': np.array(df['stress'])}
+    return output
+
+class PosteriorPredictive:
+    """
+    Will be depricated when it is integrated in probeye
+    """
+
+    def __init__(self, forward_solver, known_input_solver, parameter = None, query_kg = None):
+
+        """
+        TODO: Add option to add point estimates of parameters if forward solver is expensive
+
+        Parameters
+        ----------
+        forward_solver :
+            Defines the forward model. Check out forward_model.py to see a template for
+            the forward model definition if needed. The user will then have to derive his own
+            forward model from that base class. Examples can be found in the package
+            directory tests/integration_tests. user can also specify his/her own forward model,
+            which just takes in input.
+            Should be wrapped in a fucntion which takes in two args. arg1 = random input, arg2 = known input
+        parameter :
+            An array of samples. Currently parameters dont support a distrutbion, just MCMC samples. To get an approx.
+            for parameter dist, we need VI based posterior approx
+        known_input_solver :
+            Pass known input value to the solver in whichever format the forward_solve object needs.
+        query_kg :
+            A query script (with appropriate arguments) which outputs the sample or the stats of the inferred parameter
+            needed for the forward model
+        """
+
+        #assert (parameter == np.array), "Parameter needs to be np.ndarray"
+        #assert parameter.shape[1] == 1, "Currently support only 1 parameter"
+
+        self._forward_solver = forward_solver
+        if parameter is not None:
+            self._parameter = parameter
+        if query_kg is not None:
+            # run the query script
+            para = query_kg()
+            self._parameter = para
+        self._known_input = known_input_solver
+        self._mean = None
+        self._std = None
+        self._samples = None
+
+    def get_stats(self, samples: int) -> tuple:
+        """
+        Returns mean and s.d of the posterior predictive. Simple Monte Carlo based approximation
+
+        Parameters
+        ----------
+        samples :
+
+        Returns
+        -------
+
+        """
+        # Monte carlo step
+        output = []
+        for i in range(0, samples):
+            y = self._forward_solver(self._parameter[i], self._known_input)
+            output.append(y)
+
+        # get the posterior pred stats
+        mean = np.mean(output, axis=0)
```

Comment:
You estimate the noise as well, the posterior predictive would require an integration over the latent parameters.

Reply:
I assume the noise to be sensor noise, not model discrepancy, so dont need to include that IMO.